File: php/EvilTransform.php
<?php

namespace larryli\eviltransform;

/**
 * Package transform coordinate between earth(WGS-84) and mars in china(GCJ-02).
 * @package larryli\eviltransform
 */
class EvilTransform
{
    /**
     * outOfChina
     * @param float $lat
     * @param float $lng
     * @return bool
     */
    private static function outOfChina($lat, $lng)
    {
        if ($lng < 72.004 || $lng > 137.8347) {
            return true;
        }
        if ($lat < 0.8293 || $lat > 55.8271) {
            return true;
        }
        return false;
    }

    /**
     * transform
     * @param float $x
     * @param float $y
     * @return float[]
     */
    private static function transform($x, $y)
    {
        $xy = $x * $y;
        $absX = sqrt(abs($x));
        $d = (20.0 * sin(6.0 * $x * pi()) + 20.0 * sin(2.0 * $x * pi())) * 2.0 / 3.0;

        $lat = -100.0 + 2.0 * $x + 3.0 * $y + 0.2 * $y * $y + 0.1 * $xy + 0.2 * $absX;
        $lng = 300.0 + $x + 2.0 * $y + 0.1 * $x * $x + 0.1 * $xy + 0.1 * $absX;

        $lat += $d;
        $lng += $d;

        $lat += (20.0 * sin($y * pi()) + 40.0 * sin($y / 3.0 * pi())) * 2.0 / 3.0;
        $lng += (20.0 * sin($x * pi()) + 40.0 * sin($x / 3.0 * pi())) * 2.0 / 3.0;

        $lat += (160.0 * sin($y / 12.0 * pi()) + 320 * sin($y / 30.0 * pi())) * 2.0 / 3.0;
        $lng += (150.0 * sin($x / 12.0 * pi()) + 300.0 * sin($x / 30.0 * pi())) * 2.0 / 3.0;

        return [$lat, $lng];
    }

    /**
     * delta
     * @param float $lat
     * @param float $lng
     * @return float[] [$lat, $lng]
     */
    private static function delta($lat, $lng)
    {
        $a = 6378137.0;
        $ee = 0.00669342162296594323;
        list($dLat, $dLng) = self::transform($lng - 105.0, $lat - 35.0);
        $radLat = $lat / 180.0 * pi();
        $magic = sin($radLat);
        $magic = 1 - $ee * $magic * $magic;
        $sqrtMagic = sqrt($magic);
        $dLat = ($dLat * 180.0) / (($a * (1 - $ee)) / ($magic * $sqrtMagic) * pi());
        $dLng = ($dLng * 180.0) / ($a / $sqrtMagic * cos($radLat) * pi());
        return [$dLat, $dLng];
    }

    /**
     * WGStoGCJ convert WGS-84 coordinate(wgsLat, wgsLng) to GCJ-02 coordinate(gcjLat, gcjLng).
     * @param float $wgsLat
     * @param float $wgsLng
     * @return float[] [$gcjLat, $gcjLng]
     */
    public static function WGStoGCJ($wgsLat, $wgsLng)
    {
        if (self::outOfChina($wgsLat, $wgsLng)) {
            return [$wgsLat, $wgsLng];
        }
        list($dLat, $dLng) = self::delta($wgsLat, $wgsLng);
        return [$wgsLat + $dLat, $wgsLng + $dLng];
    }

    /**
     * GCJtoWGS convert GCJ-02 coordinate(gcjLat, gcjLng) to WGS-84 coordinate(wgsLat, wgsLng).
     *
     * The output WGS-84 coordinate's accuracy is 1m to 2m. If you want more exactly result, use GCJtoWGSExact/gcj2wgs_exact.
     * @param float $gcjLat
     * @param float $gcjLng
     * @return float[] [$wgsLat, $wgsLng]
     */
    public static function GCJtoWGS($gcjLat, $gcjLng)
    {
        if (self::outOfChina($gcjLat, $gcjLng)) {
            return [$gcjLat, $gcjLng];
        }
        list($dLat, $dLng) = self::delta($gcjLat, $gcjLng);
        return [$gcjLat - $dLat, $gcjLng - $dLng];
    }

    /**
     * GCJtoWGSExact convert GCJ-02 coordinate(gcjLat, gcjLng) to WGS-84 coordinate(wgsLat, wgsLng).
     *
     * The output WGS-84 coordinate's accuracy is less than 0.5m, but much slower than GCJtoWGS/gcj2wgs.
     * @param float $gcjLat
     * @param float $gcjLng
     * @return float[] [$wgsLat, $wgsLng]
     */
    public static function GCJtoWGSExact($gcjLat, $gcjLng)
    {
        $initDelta = 0.01;
        $threshold = 0.000001;
        // list($tmpLat, $tmpLng) = self::GCJtoWGS($gcjLat, $gcjLng);
        // list($tryLat, $tryLng) = self::WGStoGCJ($tmpLat, $tmpLng);
        // list($dLat, $dLng) = [abs($tmpLat-$tryLat), abs($tmpLng-$tryLng)];
        list($dLat, $dLng) = [$initDelta, $initDelta];
        list($mLat, $mLng) = [$gcjLat - $dLat, $gcjLng - $dLng];
        list($pLat, $pLng) = [$gcjLat + $dLat, $gcjLng + $dLng];
        list($wgsLat, $wgsLng) = [false, false];
        for ($i = 0; $i < 30; $i++) {
            list($wgsLat, $wgsLng) = [($mLat + $pLat) / 2, ($mLng + $pLng) / 2];
            list($tmpLat, $tmpLng) = self::WGStoGCJ($wgsLat, $wgsLng);
            list($dLat, $dLng) = [$tmpLat - $gcjLat, $tmpLng - $gcjLng];
            if (abs($dLat) < $threshold && abs($dLng) < $threshold) {
                // echo("i:", $i);
                return [$wgsLat, $wgsLng];
            }
            if ($dLat > 0) {
                $pLat = $wgsLat;
            } else {
                $mLat = $wgsLat;
            }
            if ($dLng > 0) {
                $pLng = $wgsLng;
            } else {
                $mLng = $wgsLng;
            }
        }
        return [$wgsLat, $wgsLng];
    }

    /**
     * Distance calculate the distance between point(latA, lngA) and point(latB, lngB), unit in meter.
     * @param float $latA lat of the point A
     * @param float $lngA lng of the point A
     * @param float $latB lat of the point B
     * @param float $lngB lng of the point B
     * @return float distance, unit in meter
     */
    public static function Distance($latA, $lngA, $latB, $lngB)
    {
        $earthR = 6371000;
        $latA *= pi() / 180;
        $lngA *= pi() / 180;
        $latB *= pi() / 180;
        $lngB *= pi() / 180;
        $x = cos($latA) * cos($latB) * cos($lngA - $lngB);
        $y = sin($latA) * sin($latB);
        $s = $x + $y;
        if ($s > 1) {
            $s = 1;
        }
        if ($s < -1) {
            $s = -1;
        }
        return acos($s) * $earthR;
    }
}

File: swift/EvilTransform.swift
//  transform.swift
//  Created by HerrKaefer on 15/6/26.

import Darwin

struct EvilTransform {
    
    static func outOfChina(#lat: Double, lng: Double) -> Bool {
        return lng < 72.004 || lng > 137.8347 || lat < 0.8293 || lat > 55.8271
    }

    static private func transform(#x: Double, y: Double) -> (lat: Double, lng: Double) {
        var lat: Double
        var lng: Double
        
        let xy = x * y
        let absX = sqrt(fabs(x))
        let d = (20.0*sin(6.0*x*M_PI) + 20.0*sin(2.0*x*M_PI)) * 2.0 / 3.0
        
        lat = -100.0 + 2.0*x + 3.0*y + 0.2*y*y + 0.1*xy + 0.2*absX
        lng = 300.0 + x + 2.0*y + 0.1*x*x + 0.1*xy + 0.1*absX
        
        lat += d
        lng += d
        
        lat += (20.0*sin(y*M_PI) + 40.0*sin(y/3.0*M_PI)) * 2.0 / 3.0
        lng += (20.0*sin(x*M_PI) + 40.0*sin(x/3.0*M_PI)) * 2.0 / 3.0
        
        lat += (160.0*sin(y/12.0*M_PI) + 320*sin(y/30.0*M_PI)) * 2.0 / 3.0
        lng += (150.0*sin(x/12.0*M_PI) + 300.0*sin(x/30.0*M_PI)) * 2.0 / 3.0
        
        return (lat, lng)
    }

    static private func delta(#lat: Double, lng: Double) -> (dLat: Double, dLng: Double) {
        var dLat: Double
        var dLng: Double
        
        let a = 6378245.0
        let ee = 0.00669342162296594323
        (dLat, dLng) = transform(x: lng-105.0, y: lat-35.0)
        let radLat = lat / 180.0 * M_PI
        var magic = sin(radLat)
        magic = 1 - ee*magic*magic
        let sqrtMagic = sqrt(magic)
        dLat = (dLat * 180.0) / ((a * (1 - ee)) / (magic * sqrtMagic) * M_PI)
        dLng = (dLng * 180.0) / (a / sqrtMagic * cos(radLat) * M_PI)
        return (dLat, dLng)
    }

    static func wgs2gcj(#wgsLat: Double, wgsLng: Double) -> (gcjLat: Double, gcjLng: Double) {
        if outOfChina(lat: wgsLat, lng: wgsLng) {
            return (wgsLat, wgsLng)
        }
        var dCoordinate = delta(lat: wgsLat, lng: wgsLng)
        return (wgsLat + dCoordinate.dLat, wgsLng + dCoordinate.dLng)
    }

    static func gcj2wgs(#gcjLat: Double, gcjLng: Double) -> (wgsLat: Double, wgsLng: Double) {
        if outOfChina(lat: gcjLat, lng: gcjLng) {
            return (gcjLat, gcjLng)
        }
        var dCoordinate = delta(lat: gcjLat, lng: gcjLng)
        return (gcjLat - dCoordinate.dLat, gcjLng - dCoordinate.dLng)
    }

    // nIter=2: centimeter precision, nIter=5: double precision
    static func gcj2wgs_exact(#gcjLat: Double, gcjLng: Double, nIter: Int = 2) -> (wgsLat: Double, wgsLng: Double) {
        if outOfChina(lat: gcjLat, lng: gcjLng) {
            return (gcjLat, gcjLng)
        }
        var wgsLat = gcjLat
        var wgsLng = gcjLng
        for _ in 0..<nIter {
            var dCoordinate = delta(lat: wgsLat, lng: wgsLng)
            wgsLat = gcjLat - dCoordinate.dLat
            wgsLng = gcjLng - dCoordinate.dLng
        }
        return (wgsLat, wgsLng)
    }

    // 1 - cos(x) == 2 sin^2(x/2)
    static private func oneMinusCos(x: Double) -> Double
    {
        let s = sin(x/2)
        return s*s*2
    }

    static func distance(#latA: Double, lngA: Double, latB: Double, lngB: Double) -> Double {
        let earthR: Double = 6371000
        let dLatA = latA * M_PI / 180
        let dLatB = latB * M_PI / 180
        let dLngA = lngA * M_PI / 180
        let dLngB = lngB * M_PI / 180
        return 2 * earthR * asin(sqrt(oneMinusCos(dLatA-dLatB) + cos(dLatA)*cos(dLatB)*(oneMinusCos(dLngA - dLngB)))/M_SQRT2)
    }
}

func testEvilTransform() {
    let testCoordinates = [
        // wgsLat, wgsLng, gcjLat, gcjLng
        [31.1774276, 121.5272106, 31.17530398364597, 121.531541859215], // shanghai
        [22.543847, 113.912316, 22.540796131694766, 113.9171764808363], // shenzhen
        [39.911954, 116.377817, 39.91334545536069, 116.38404722455657], // beijing
    ]
    
    for coord in testCoordinates {
        let wgsLat = coord[0]
        let wgsLng = coord[1]
        let gcjLat = coord[2]
        let gcjLng = coord[3]
    
        println("wgs2gcj:")
        var gcjCoord = EvilTransform.wgs2gcj(wgsLat: wgsLat, wgsLng: wgsLng)
        println("gcj         : \(gcjLat), \(gcjLng)")
        println("gcj computed: \(gcjCoord.gcjLat), \(gcjCoord.gcjLng)")

        println("gcj2wgs:")
        var wgsCoord = EvilTransform.gcj2wgs(gcjLat: gcjLat, gcjLng: gcjLng)
        var d = EvilTransform.distance(latA: wgsCoord.wgsLat, lngA: wgsCoord.wgsLng, latB: wgsLat, lngB: wgsLng)
        if d > 5 {
            println("gcj2wgs test \(coord): distance \(d)")
        }
    
        println("gcj2wgs_exact:")
        wgsCoord = EvilTransform.gcj2wgs_exact(gcjLat: gcjLat, gcjLng: gcjLng)
        d = EvilTransform.distance(latA: wgsCoord.wgsLat, lngA: wgsCoord.wgsLng, latB: wgsLat, lngB: wgsLng)
        if d > 0.5 {
            println("gcj2wgs_exact test \(coord): distance \(d)")
        }
    }
}
